### PR TITLE
Feat/improved tools

### PR DESCRIPTION
**Pull request by richardblythman**
1. Add improved scraping to tools
2. Add RAG tool

### CODE REVIEW

**Comment by richardblythman** [comment]
@jmoreira-valory @Adamantios this is the branch that we need a review on.

File: tools/prediction_request_rag/prediction_request_rag.py [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+        results = result_dict
+        results = json.dumps(results)
+        if counter_callback is not None:
+            counter_callback(
+                input_tokens=response.usage.prompt_tokens,
+                output_tokens=response.usage.completion_tokens,
+                model=engine,
+            )
+            return results, prediction_prompt, counter_callback
+        return results, prediction_prompt, None
```
Comment (0xArdi) [inline tools/prediction_request_rag/prediction_request_rag.py:572]: The response is expected to be the same across all tools. It has to be (result: str, prompt: str, transaction: Optional[Dict], counter_callback: TokenCounterCallback)

Example: https://github.com/valory-xyz/mech/blob/main/packages/valory/customs/prediction_request/prediction_request.py#L441

File: tools/prediction_request_rag/prediction_request_rag.py [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+    queries = Queries.from_response(response)
+
+    # append the user's question to the list of queries
+    queries.queries.append(prompt)
+
+    if counter_callback:
+        counter_callback(
+            input_tokens=response.usage.prompt_tokens,
+            output_tokens=response.usage.completion_tokens,
+            model=engine,
```
[the remark this answers is not in the document]
Comment (moarshy) [reply tools/prediction_request_rag/prediction_request_rag.py:236]: done

File: tools/prediction_request_rag/prediction_request_rag.py (hunk 1 above)
Comment (moarshy) [reply tools/prediction_request_rag/prediction_request_rag.py]: done

**Review by jmoreira-valory — CHANGES_REQUESTED** [review]
Please merge main branch here and structure the tools under 
`packages/<vendor>/customs/<custom_package_name>`

**Review by jmoreira-valory — APPROVED** [review]

**Comment by jmoreira-valory** [comment]
Note: This PR has been restructured to be merge ready with the new structure of the packages directory:

1. Main branch has been merged here
2. Each tool has been moved to their corresponding location under `packages/vendor/customs/tool`
3. The source code of the tools has been updated to the latest changes in this branch at commit https://github.com/valory-xyz/mech/pull/164/commits/da70f350734189e929ddb2019477ffd3af384b68 (already merged changes in main will therefore be hidden).
4. Hashes have been updated.

File: packages/napthaai/customs/prediction_request_rag/prediction_request_rag.py [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+        if counter_callback is not None:
+            counter_callback(
+                input_tokens=response.usage.prompt_tokens,
+                output_tokens=response.usage.completion_tokens,
+                model=engine,
+                token_counter=count_tokens,
+
+            )
+            return results, prediction_prompt, None, counter_callback
+        return results, prediction_prompt, None, None
```
Comment (Adamantios) [inline packages/napthaai/customs/prediction_request_rag/prediction_request_rag.py:582]:
```suggestion
        return results, prediction_prompt, None, counter_callback

```

File: packages/nickcom007/customs/prediction_request_sme/prediction_request_sme.py [hunk 4]
```diff
@@ -226,13 +231,21 @@ def extract_text(
...
-    for script in soup(["script", "style"]):
-        script.extract()
-    text = soup.get_text()
-    lines = (line.strip() for line in text.splitlines())
-    chunks = (phrase.strip() for line in lines for phrase in line.split("  "))
-    text = "\n".join(chunk for chunk in chunks if chunk)
+    text = Document(html).summary()
+
+    # use html2text to convert HTML to markdown
+    text
```
Comment (Adamantios) [inline packages/nickcom007/customs/prediction_request_sme/prediction_request_sme.py:237]: Can you explain this line?

File: packages/valory/customs/prediction_request/prediction_request.py [hunk 5]
```diff
@@ -298,17 +306,26 @@ def fetch_additional_information(
...
         counter_callback(
             input_tokens=response.usage.prompt_tokens,
             output_tokens=response.usage.completion_tokens,
             model=engine,
             token_counter=count_tokens,
         )
-        return "\n".join(["- " + text for text in texts]), counter_callback
-    return "\n".join(["- " + text for text in texts]), None
+        return additional_information, counter_callback
+    return additional_information, None
```
Comment (Adamantios) [inline packages/valory/customs/prediction_request/prediction_request.py:328]:
```suggestion
    return additional_information, counter_callback
```

File: packages/valory/customs/prediction_request_claude/prediction_request_claude.py [hunk 6]
```diff
@@ -167,17 +167,24 @@ def get_urls_from_queries(queries: List[str], api_key: str, engine: str) -> List
 
 def extract_text(
     html: str,
-    num_words: int = 300,  # TODO: summarise using LLM instead of limit
+    num_words: Optional[int] = 300,
```
Comment (Adamantios) [inline packages/valory/customs/prediction_request_claude/prediction_request_claude.py:170]: Maybe default to `None`?

File: packages/valory/customs/prediction_request_claude/prediction_request_claude.py [hunk 7]
```diff
@@ -252,17 +260,26 @@ def fetch_additional_information(
...
         counter_callback(
             model=engine,
             input_prompt=url_query_prompt,
             output_tokens=40,
             token_counter=count_tokens,
         )
-        return "\n".join(["- " + text for text in texts]), counter_callback
-    return "\n".join(["- " + text for text in texts]), None
+        return additional_information, counter_callback
+    return additional_information, None
```
Comment (Adamantios) [inline packages/valory/customs/prediction_request_claude/prediction_request_claude.py:282]:
```suggestion
    return additional_information, counter_callback
```

File: packages/nickcom007/customs/prediction_request_sme/prediction_request_sme.py (hunk 4 above)
Comment (moarshy) [reply packages/nickcom007/customs/prediction_request_sme/prediction_request_sme.py:237]: This seem to be missing something. it should have been

`text = md(text, heading_style="ATX")`

File: packages/valory/customs/prediction_request_claude/prediction_request_claude.py [hunk 8]
```diff
@@ -167,17 +167,24 @@ def get_urls_from_queries(queries: List[str], api_key: str, engine: str) -> List
...
+    text = md(text, heading_style="ATX")
+
+    if text is None:
+        return ""
+
+    # remove newlines and extra spaces
+    text = " ".join(text.split())
+
+    if num_words:
+        return " ".join(text.split()[:num_words])
```
[the remark this answers is not in the document]
Comment (moarshy) [reply packages/valory/customs/prediction_request_claude/prediction_request_claude.py:185]: no, maybe can bring the first split below.

File: packages/valory/customs/prediction_request_claude/prediction_request_claude.py (hunk 6 above)
Comment (moarshy) [reply packages/valory/customs/prediction_request_claude/prediction_request_claude.py:170]: ok

**Review by Adamantios — APPROVED** [review]

**Review by 0xArdi — APPROVED** [review]

**Pull request merged**